### DISCUSSION TITLE
IPSec from Home Office using IPv6 without NAT

### DISCUSSION THREAD

**Discussion by bjoern30** [strongswan/strongswan#2139] [General]
We are using a software solution, which integrates a strongswan client, for an IPSec Tunnel to a device using iked on openbsd. In this case IKE version 2 is used. 
The clients connect to the gateway through internet using IPv4 or IPv6 (configured as Dualstack). 
Using IPv4 network address translation is done in every case, so NAT traversal is done always. 
Using IPv6 NAT is done only in very few working cases. 

The access with IPv4 is working well. 
While using IPv6 the connection / IPSec tunnel is closed after approx. five minutes. 
I have analyzed the behavior: 
-	1 The client / initiator sends a udp datagram on port 500 to the gateway / server. 
-	2 The server answers this udp datagram on port 500. 
          _In this part all necessary informations for doing a NAT detection are exchanged._
 
-	3 The client sends a udp datagram on port 4500 to the gateway / server. 
-	4 The server answers this udp datagram on port 4500. 
         _With these packets the informations for doing a NAT detection are exchanged again._ 
-	The client detects that NO NAT is used.
-	The client send further on packets using ESP to the gateway.
         _The home router is configured as default to allow outgoing connections and to block incoming connections (except answer packets)._ 
-	The home router between client and server / gateway closes the open udp port due inactivity after between 30 and 300 seconds. 
-	After 300 and a few seconds the gateway sends an informational responder request on udp port 4500 to the client / initiator. 
-	This datagram is blocked on the home router due the timed out packet filter state. The home router sends an icmp destination unreachable administratively prohibited to the gateway.
In some cases the router blocks without sending an icmp destination unreachable message.
-	The gateway / server / responder sends the informational responder request on udp port 4500 five more times to the client / initiator. This packets are also blocked by the packet filter. 
-	The gateway now thinks the initiator is down / not responding (although the initiator is sending lots of ESP packets at the same time) and is shutting down the ipsec tunnel. 

I have asked several persons to this and got the following answers and questions: 
-	Mobile clients must always use NAT-T. The client must be started using “forceencaps=yes”.
        I am not sure about that. Why does the client NAT detection, if only NAT-Traversal is allowed? 

-	If there was no NAT, why did your initiator switch from port 500 to port 4500?

The initiator may detect after datagrams 1 and 2, that no NAT is used. Why is the initiator sending the udp datagram 3 on port 4500.

In my opinion there are two solutions:
-	Enforce the client to use NAT-T. This might be the fastest workaround with maybe bad side effects.

-	Change the software, that the packets three and four actual send on udp/4500 are sent in the future via ESP.

**Comment by tobiasbrunner** [comment]
> * Mobile clients must always use NAT-T. The client must be started using “forceencaps=yes”.
>       I am not sure about that. Why does the client NAT detection, if only NAT-Traversal is allowed?

It's definitely not a must. But it might work a lot more stable with UDP-encapsulation as that doesn't require plain ESP packets (which some ISPs might block) and uses the same UDP-"channel" already used by the IKE packets (e.g. DPDs and NAT keepalives keep that open).

> * If there was no NAT, why did your initiator switch from port 500 to port 4500?

[MOBIKE](https://docs.strongswan.org/docs/5.9/features/mobike.html)


> * Enforce the client to use NAT-T. This might be the fastest workaround with maybe bad side effects.

What bad side effects are you thinking about?

> * Change the software, that the packets three and four actual send on udp/4500 are sent in the future via ESP.

You want to send IKE packets via ESP? Without having an IPsec SA negotiated yet? Not only because of the latter that's not an option.

What you could do is force the client to send NAT keepalives via IKE ports even if no NAT is detected (requires code changes). That way the packets from the server should get through as the firewall should keep the port open. Or make the client send DPDs relatively often (although that might close the SA if the connectivity is interrupted for a while, and the default retransmission times won't be suitable for this use case).

**Comment by bjoern30** [comment]
Thanks for your answer.

Before I've opened this discussion I have searched some time and found some websites containing the information, that NAT-T could cause in some situations problems, you don't have with ESP. Unfortunately this problems were not described in Detail. 
Therefore we decided to support the normal use with ESP and the use of NAT-T, if NAT is used. 

- You want to send IKE packets via ESP?
My view was looking from the packet filter. While there is a constant transmission, the packet filter state will not time out. Sending on multiple ports is always dangerous for timing out ports. 

If a keepalive will be sent from the client to the server the repeating time must the coordinated between developers of ipsec, firewalls, packet filter and so on. A client must send a keepalive packet short before the responder sends the informational responder request.
This might be difficult to coordinate between all stakeholders.

In my opinion you are aware of the problem. I have informed also the authors of the rfc7296 (ikev2) about this issue. 

At first, we discussed with our software developer to enforce NAT-T using “forceencaps=yes”. This might be a good solution at first. 
On a longer sight it would be good if the communication with ESP + UDP/4500 will work over firewalls in networks without NAT over firewalls like in home routers.

**Reply by tobiasbrunner** [reply]
> My view was looking from the packet filter. While there is a constant transmission, the packet filter state will not time out. Sending on multiple ports is always dangerous for timing out ports.

ESP does not have any ports, it's a separate IP protocol.

> If a keepalive will be sent from the client to the server the repeating time must the coordinated between developers of ipsec, firewalls, packet filter and so on. A client must send a keepalive packet short before the responder sends the informational responder request.
> This might be difficult to coordinate between all stakeholders.

There is no coordination about this. NAT keepalives are simply sent every 20 seconds, by default, if a host detects it's behind a NAT (they are simple UDP packets with a one-byte payload).

**Reply by tobiasbrunner** [reply]
By the way, you might also be interested in #1759.

**Comment by bjoern30** [comment]
OK, thanks

**Comment by bjoern30** [comment]
Yeah, that seems to be the same. While querying before opening the discussion I did not find this discussion. Wrong key words ;-)

Von: Tobias Brunner ***@***.***>
Gesendet: Dienstag, 5. März 2024 15:45
An: strongswan/strongswan ***@***.***>
Cc: bjoern30 ***@***.***>; State change ***@***.***>
Betreff: Re: [strongswan/strongswan] IPSec from Home Office using IPv6 without NAT (Discussion #2139)


By the way, you might also be interested in #1759<https://github.com/strongswan/strongswan/issues/1759>.
